apps/emqx_dashboard/src/emqx_dashboard_login_lock.erl

```diff
@@ -0,0 +1,280 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2025 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-module(emqx_dashboard_login_lock).
+
+-include("emqx_dashboard.hrl").
+
+-include_lib("emqx/include/logger.hrl").
+
+-behaviour(gen_server).
+
+-export([
+    register_unsuccessful_login/1,
+    verify/1,
+    reset/1,
+    is_login_locked_error/1
+]).
+
+-export([
+    create_tables/0,
+    start_link/0
+]).
+
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2
+]).
+
+%% For testing/debugging
+-export([
+    cleanup_all/0,
+    info/0
+]).
+
+-type time_s() :: non_neg_integer().
+-type time_us() :: non_neg_integer().
+-type username() :: binary().
+
+-record(login_lock, {
+    username :: username() | '_',
+    locked_until :: time_s() | '_'
+}).
+
+%% NOTE
+%% We ignore the fact that we may lose some attempts
+%% if they happen in the same microsecond.
+-record(login_attempts, {
+    key :: {username(), time_us()} | {'_', '_'} | '_',
+    extra :: map() | '_'
+}).
+
+-define(TAB_ATTEMPTS, login_attempts).
+-define(TAB_LOCK, login_lock).
+
+-ifdef(TEST).
+-define(CLEANUP_INTERVAL, 100).
+-else.
+-define(CLEANUP_INTERVAL, timer:minutes(30)).
+-endif.
+
+-define(ETS_CHUNK_SIZE, 100).
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+-spec register_unsuccessful_login(username()) -> ok.
+register_unsuccessful_login(Username) ->
+    NowUs = now_us(),
+    Key = {Username, NowUs},
+    Attempt = #login_attempts{key = Key, extra = #{}},
+    max_attempts_reached(Username, NowUs) andalso lock(Username, NowUs),
+    ok = mria:dirty_write(?TAB_ATTEMPTS, Attempt).
```

Comment:
Does this write happen even if the username is locked?

If so, this feature becomes an easily explaitable DOS vector: since unsuccessful login attempts are stored in a `disc_copies` table (a.k.a. ets with some disc log persistence), every attempt will create a record in RAM. Fire them fast enough, and you crash the server.

Reply:
Does this write happen even if the username is locked?

No, the record is added only on failed password check, and in case of a present lock, the password check is not performed.

Reply:
What about non-existent usernames?